pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -566,61 +593,75 @@ func buildHTTPConnectionManager(mesh *meshconfig.MeshConfig, httpOpts *httpListe
 
 // buildListener builds and initializes a Listener proto based on the provided opts. It does not set any filters.
 func buildListener(opts buildListenerOpts) *xdsapi.Listener {
-	var filterChainMatch *listener.FilterChainMatch
-	if len(opts.sniHosts) > 0 {
-		filterChainMatch = &listener.FilterChainMatch{
-			SniDomains: opts.sniHosts,
+	filterChains := make([]listener.FilterChain, 0, len(opts.filterChainOpts))
+	for _, chain := range opts.filterChainOpts {
+		match := &listener.FilterChainMatch{}
+		if len(chain.sniHosts) > 0 {
+			match.SniDomains = chain.sniHosts
 		}
+		filterChains = append(filterChains, listener.FilterChain{
+			FilterChainMatch: match,
+			TlsContext:       chain.tlsContext,
+		})
 	}
 
 	var deprecatedV1 *xdsapi.Listener_DeprecatedV1
 	if !opts.bindToPort {
 		deprecatedV1 = &xdsapi.Listener_DeprecatedV1{
-			BindToPort: &google_protobuf.BoolValue{
-				Value: false,
-			},
+			BindToPort: boolFalse,
 		}
 	}
-	var protoPrefix string
+	var protocolPrefix string
 	if opts.protocol.IsHTTP() {
-		protoPrefix = "http"
+		protocolPrefix = "http"
 	} else {
-		protoPrefix = "tcp"
+		protocolPrefix = "tcp"
```

Comment:
put this block in helper for better readability?

Reply:
Done